lib/notifications/notification.ex

```diff
@@ -66,26 +68,54 @@ defmodule Notifications.Notification do
     changeset =
       DbNotification.changeset(%DbNotification{}, Map.from_struct(notification_without_id))
 
-    db_record = insert!(changeset, on_conflict: :nothing)
+    {:ok, notification_with_id} =
+      Skate.Repo.transaction(fn ->
+        db_record = insert!(changeset, on_conflict: :nothing)
 
-    db_record =
-      if db_record.id do
-        notification_with_id = %__MODULE__{notification_without_id | id: db_record.id}
-        log_creation(notification_with_id)
-        db_record
-      else
-        identifying_fields =
-          notification_without_id
-          |> Map.take([:start_time, :end_time, :block_id, :service_id, :reason])
-          |> Map.to_list()
+        db_record =
+          if db_record.id do
+            notification_with_id = %__MODULE__{notification_without_id | id: db_record.id}
+            log_creation(notification_with_id)
+            link_notification_to_users(notification_with_id)
+            db_record
+          else
+            identifying_fields =
+              notification_without_id
+              |> Map.take([:start_time, :end_time, :block_id, :service_id, :reason])
+              |> Map.to_list()
 
-        Skate.Repo.one(from(DbNotification, where: ^identifying_fields))
-      end
+            Skate.Repo.one(from(DbNotification, where: ^identifying_fields))
+          end
 
-    %__MODULE__{notification_without_id | id: db_record.id}
+        %__MODULE__{notification_without_id | id: db_record.id}
+      end)
+
+    notification_with_id
   end
 
   defp log_creation(notification) do
     Logger.warn("Notification created new_notification=#{inspect(notification)}")
   end
+
+  defp link_notification_to_users(notification) do
+    now = NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second)
+
+    user_ids = User.user_ids_for_route_ids(notification.route_ids)
+
+    notification_user_maps =
+      Enum.map(user_ids, fn user_id ->
+        %{
+          notification_id: notification.id,
+          user_id: user_id,
+          state: :unread,
+          inserted_at: now,
+          updated_at: now
+        }
+      end)
+
+    Skate.Repo.insert_all(
+      DbNotificationUser,
+      notification_user_maps
+    )
```

Comment:
So, one other thing we could consider is actually using `users_for_route_ids_query` directly and building that into a big SQL query to insert the relevant rows, without having the relevant logic to make `notification_user_maps` on the Elixir side. That said, that's a lot of rewriting stuff by trial and error in weird Ecto query DSL form and / or fragments and we're not really trying to solve for a race condition or anything, so I think it's fine as-is.